### ISSUE TITLE
Freezes everything if gpg key decryption fails when trying to interact with remote password repository, of which password is stored in pass

### ISSUE THREAD

**Issue by AlexAegis** [languitar/pass-git-helper#17]
This is printed on both fails, but after the second my terminal emulator (and everything else) hangs. I have to change tty-s to kill the terminal window.

```sh
gpg: decryption failed: No secret key
Traceback (most recent call last):
  File "/usr/bin/pass-git-helper", line 11, in <module>
    load_entry_point('pass-git-helper==1.1.0', 'console_scripts', 'pass-git-helper')()
  File "/usr/lib/python3.8/site-packages/passgithelper.py", line 429, in main
    get_password(request, mapping)
  File "/usr/lib/python3.8/site-packages/passgithelper.py", line 371, in get_password
    output = subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['pass', 'show', 'git/github']' returned non-zero exit status 2.
```

It works when I put my password in.

**Comment by languitar** [comment]
I can't do much here apart from exiting the script in case a password cannot be retrieved and at least on terminal, git handles this gracefully by resorting to a manual password prompt:
```
❯ git clone https://gitlab.com/johannes.wienke/infrastructure-doc.git
Cloning into 'infrastructure-doc'...
gpg: decryption failed: No secret key
Traceback (most recent call last):
  File "/usr/bin/pass-git-helper", line 11, in <module>
    load_entry_point('pass-git-helper==1.1.0', 'console_scripts', 'pass-git-helper')()
  File "/usr/lib/python3.8/site-packages/passgithelper.py", line 429, in main
    get_password(request, mapping)
  File "/usr/lib/python3.8/site-packages/passgithelper.py", line 371, in get_password
    output = subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['pass', 'show', 'dev/gitlab.com']' returned non-zero exit status 2.
Username for 'https://gitlab.com': 
```
Can you try this again on a plain terminal?

**Comment by languitar** [comment]
It seems the [documentation about credential helpers](https://git-scm.com/docs/api-credentials) got a lot better since I last looked at it.

After your test on plain terminal, if only vscode is affected and hangs due to the input prompt, maybe an added option for the following behavior could help you:

> If a helper outputs a quit attribute with a value of true or 1, no further helpers will be consulted, nor will the user be prompted (if no credential has been provided, the operation will then fail).

However, we can also test this first then with some bash magic.

**Comment by AlexAegis** [comment]
It doesn't matter what kind of terminal I use, everything freezes, other windows too, I wouldn't have to switch to another TTY if only one window would freeze.

But I wasn't clear in the issue. This only happens when I try to interact with the remote git repo the passwords are in. (`pass git pull` for example)
The password store is in a private repo.

If I just want to print something from the local repo, then leave the password prompt empty, or enter a wrong password thrice (the gpg prompt, doesn't matter if in the terminal or using a pinentry program. `pinentry-program /usr/bin/pinentry-qt`), it just exits.

> gpg: decryption failed: No secret key.

But when using `pass git pull` the aforementioned thing happens.
I think it's some kind of infinite loop, trying to get the password from the password store to pull from the remote password store which has it's password in the password store. It probably block everything else, trying to execute that.

**Comment by languitar** [comment]
Alright, all of this sounds pretty suspicious.

First of all a meta comment: pass-git-helper should only be used for cases where it is not possible to use better authentication schemes such as SSH keys. Especially for your private repo containing the password store files I would always recommend to use SSH keys to have a safe authentication mechanism (SSH keys are some light form of two-factor authentication because you need the key file and the passphrase). Moreover, that also avoids the chicken and egg problem when bootstrapping a new machine.

Regarding the real problem: one suspicion that I have is that something configured in your shell environment tries to interact with the password store (maybe for completion) and this gets blocked by the ongoing operation on the repository. To verify this, one thing could be to have another terminal open in parallel while being in the otherwise blocked state that uses a shell such as `sh` with no completion and further configuration. Is that shell still usable?

Another thing we could do: if one of your shells is still usable when another is blocked, you could try to start `strace <yourshell>` in there. That way, we see with which files or other programs the shell interacts before getting stuck.

**Comment by AlexAegis** [comment]
First comment: That is a very good advice! It's mandatory anyway for two-way authentication.

And second:
from `strace` I didn't really learned anything but just by checking `htop` I saw something odd in the process tree. Turns out, that if I fail to let gpg do its thing, the next thing that will run run is executing `SSH_ASKPASS`, which was misconfigured in my setup. If I unset it it fails like it should.

Sorry to bother you, I'll close this now.
